lightning/src/offers/flow.rs

```diff
@@ -1082,4 +1133,317 @@ where
+	/// [`ServeStaticInvoice`] onion messages in response if we've built a new async receive offer and
+	/// need the corresponding [`StaticInvoice`] to be persisted by the static invoice server.
+	///
+	/// Returns `None` if we have enough offers cached already, verification of `message` fails, or we
+	/// fail to create blinded paths.
+	#[cfg(async_payments)]
+	pub(crate) fn handle_offer_paths<ES: Deref, R: Deref>(
+		&self, message: OfferPaths, context: AsyncPaymentsContext, responder: Responder,
+		peers: Vec<MessageForwardNode>, usable_channels: Vec<ChannelDetails>, entropy: ES,
+		router: R,
+	) -> Option<(ServeStaticInvoice, MessageContext)>
+	where
+		ES::Target: EntropySource,
+		R::Target: Router,
+	{
+		let duration_since_epoch = self.duration_since_epoch();
+		match context {
+			AsyncPaymentsContext::OfferPaths { path_absolute_expiry } => {
+				if duration_since_epoch > path_absolute_expiry {
+					return None;
+				}
+			},
+			_ => return None,
+		}
+
+		{
+			// Only respond with `ServeStaticInvoice` if we actually need a new offer built.
+			let mut cache = self.async_receive_offer_cache.lock().unwrap();
+			cache.prune_expired_offers(duration_since_epoch, false);
+			if !cache.should_build_offer_with_paths(
+				&message.paths[..],
+				message.paths_absolute_expiry,
+				duration_since_epoch,
+			) {
+				return None;
+			}
+		}
+
+		let (mut offer_builder, offer_nonce) =
+			match self.create_async_receive_offer_builder(&*entropy, message.paths) {
+				Ok((builder, nonce)) => (builder, nonce),
+				Err(_) => return None, // Only reachable if OfferPaths::paths is empty
+			};
+		if let Some(paths_absolute_expiry) = message.paths_absolute_expiry {
+			offer_builder =
+				offer_builder.absolute_expiry(Duration::from_secs(paths_absolute_expiry));
+		}
+		let (offer_id, offer) = match offer_builder.build() {
+			Ok(offer) => (offer.id(), offer),
+			Err(_) => {
+				debug_assert!(false);
+				return None;
+			},
+		};
+
+		let (invoice, forward_invoice_request_path) = match self.create_static_invoice_for_server(
+			&offer,
+			offer_nonce,
+			peers,
+			usable_channels,
+			&*entropy,
+			router,
+		) {
+			Ok(res) => res,
+			Err(()) => return None,
+		};
+
+		let res = self.async_receive_offer_cache.lock().unwrap().cache_pending_offer(
+			offer,
+			message.paths_absolute_expiry,
+			offer_nonce,
+			responder,
+			duration_since_epoch,
+		);
+
+		let invoice_slot = match res {
+			Ok(idx) => idx,
+			Err(()) => return None,
+		};
+
+		let reply_path_context = {
+			let path_absolute_expiry =
+				duration_since_epoch.saturating_add(TEMP_REPLY_PATH_RELATIVE_EXPIRY);
+			MessageContext::AsyncPayments(AsyncPaymentsContext::StaticInvoicePersisted {
+				offer_id,
+				path_absolute_expiry,
+			})
+		};
+
+		let serve_invoice_message =
+			ServeStaticInvoice { invoice, forward_invoice_request_path, invoice_slot };
+		Some((serve_invoice_message, reply_path_context))
```

Comment:
This is probably how the framework is set up, but it seems somewhat inconsistent that we queue onion messages in flow.rs and also send responses in messenger.rs